metabase/metabase
# fix(sdk): support React 17 backwards compatibility

Closes https://github.com/metabase/metabase/issues/45713

### Description

Dropping support for React 17 means that most of the SDK customers can't use the SDK. The incompatibility with React 17 is caused by:

- [Upgrading react-redux to v9](https://github.com/metabase/metabase/pull/44339) - it uses useSyncExternalStore without any compatibility shims, which is only available in React 18. RTK 2.0 seems to [require](https://redux.js.org/usage/migrations/migrating-rtk-2) react-redux v9. We need to monkey-patch `React.useSyncExternalStore` so that it works for React 17.
- Use of `createRoot` in `TableInteractive`, which is not supported in React 17
- Back button crashing because of it not returning `null`

### How to verify

- Setup a test application using React 17 and the Embedding SDK
  - Alternatively, you could use the `test-react-17` branch of Shoppy (shown in the demo video below), but the CSS are pretty broken because of downgrading from Mantine v7 to Mantine v6. This is an issue with Shoppy's use of Mantine v7, not with the SDK.
- The SDK should still work fine. Try the following code paths:
  - Static Questions
  - Interactive Questions
  - Drill Downs to Interactive Questions (tables) 
  - Interactive Dashboards

### Demo

<div>
    <a href="https://www.loom.com/share/c30f61fbc94d42d5ae975a15446f022f">
      <p>React 17 backwards compatibility - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/c30f61fbc94d42d5ae975a15446f022f">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/c30f61fbc94d42d5ae975a15446f022f-3e80a821b599d1cd-full-play.gif">
    </a>
  </div>

### Notes

- Will add sanity checks specifically for React 17 in a separate PR (see https://github.com/metabase/metabase/issues/45714), it'll take time to setup though.
